### PR TITLE
Select only new senses

### PR DESCRIPTION
**Pull request by micgro42**
This still needs to be actually tested

### CODE REVIEW

**Comment by micgro42** [comment]
I tried it out locally by switching to sqlite, mocking the user/session and with that it _seems_ to work. Though I'm not sure if we have to run some actual commands on the server, maybe with prisma, because we're changing the database schema?

**Comment by micgro42** [comment]
So, it should be possible to apply the migrations on toolforge with the prisma CLI. I _think_ this is good to go. After all, this whole thing is a prototype that is not yet announced or working anywhere anyway.

File: server/sparql-queries.js [hunk 1]
```diff
@@ -4,11 +4,12 @@ function sanitize( str ){
 }
 
 module.exports = {
-    itemlessSenses: ( languageCode, languageQid ) => {
+    itemlessSenses: ( languageCode, languageQid, senseIdsToSkip ) => {
         const nounQid = 'Q1084';
         const itemForThisSensePid = 'P5137';
         const translationPid = 'P5972';
         const numberOfSensesPerRequest = 10;
+        const prefixedSensIdsToSkip = senseIdsToSkip.map( senseId => `wd:${senseId}` );
```
Comment (lucaswerkmeister) [inline server/sparql-queries.js]: typo: prefixedSens**e**IdsToSkip

File: server/sparql-queries.js [hunk 2]
```diff
@@ -21,6 +22,7 @@ module.exports = {
             FILTER(LANG(?gloss) = "${languageCode}")
             MINUS { ?senseId wdt:${itemForThisSensePid} [] }
             MINUS { ?senseId wdt:${translationPid} [] }
+            FILTER(?senseId NOT IN (${prefixedSensIdsToSkip.join(', ')}))
```
Comment (lucaswerkmeister) [inline server/sparql-queries.js]: The query will look slightly weird if `senseIdsToSkip` is empty, but it looks like `FILTER(?x NOT IN ())` is valid SPARQL (or at least WDQS accepts it).

**Review by lucaswerkmeister — COMMENTED** [review]
Looks okay to me (apart from a typo), but I’m a bit worried about how it will scale. How many rejected sense IDs do we expect to have, if we’re adding all of them to a query?

Can we maybe limit the `senseIdsToSkip` to only the ones that are relevant to the query (probably using the same language item ID)?

File: server/sparql-queries.js (hunk 1 above)
Comment (micgro42) [reply server/sparql-queries.js]: Fixed in force-pushed commit

**Review by lucaswerkmeister — APPROVED** [review]
LGTM for now, could be improved further of coruse

**Comment by micgro42** [comment]
Ok, _now_ it uses the correct value from the user-object to store in the database: the userId. Though, I am a bit skeptical, because it tells me that the userId of my private account "Zvpunry" is `6619`, which seems ... _low_? Not sure that this id is actually stable?

Also, I also added storing the language of the sense to this PR right away. The reason is that adding the language can't be done in a good way without either loosing the entire db or looking up the language of each sense to backfill each row.

**Comment by lucaswerkmeister** [comment]
Apparently 6619 is your CentralAuth ID, so I guess you enabled SUL for your account fairly early? https://www.wikidata.org/w/api.php?action=query&format=json&list=allusers&aufrom=Zvpunry&auto=Zvpunry&auprop=centralids

**Comment by micgro42** [comment]
Is there a good way to go from CentralAuth ID to username?

**Comment by lucaswerkmeister** [comment]
[meta=globaluserinfo](https://www.wikidata.org/w/api.php?action=query&format=json&meta=globaluserinfo&guiid=6619), apparently (though I had to find that in [Extension:CentralAuth/API](https://www.mediawiki.org/wiki/Extension:CentralAuth/API))

**Comment by codders** [comment]
Closing this PR as we are deprecating the tool and archiving the repository.

**Pull request closed without merging**